### PR TITLE
Separate clean, reset, and destroy logic

### LINKED ISSUE #26: Refactor components destroy function
Currently, we have one destroy method for components that accepts parameters determining whether the component should be reset or cleaned as well. In order to make this easier to reuse and implement, we should introduce three methods: destroy, reset, and clean. Each of these methods then only accepts the dry-run parameter. The component calls destroy is then responsible to call in an order the applicable methods: destroy, reset, and clean.
One thing to note: if we split reset and clean, then resetters will be deployed twice with different configs. So maybe we should just split out and destroy.

### PR DESCRIPTION
**Pull request by raminqaf**
Closes #26

### CODE REVIEW

File: kpops/cli/main.py [hunk 1]
```diff
@@ -191,7 +192,9 @@ def run_destroy_clean_reset(
     )
     for component in reversed(get_steps_to_apply(pipeline, steps)):
         log_action("Destroy", component)
-        component.destroy(dry_run=dry_run, clean=clean, delete_outputs=delete_outputs)
+        component.destroy(dry_run=dry_run)
+        if clean and isinstance(component, KafkaApp):
```
Comment (philipp94831) [inline kpops/cli/main.py]: Why do you check for is instance? I would just make the default implementation do nothing and so far only KafkaApp overrides it

Comment (philipp94831) [reply kpops/cli/main.py]: haha, you can read my mind 😂

File: kpops/components/base_components/kafka_app.py [hunk 2]
```diff
@@ -63,23 +66,24 @@ def deploy(self, dry_run: bool):
...
-            # producer clean up job will delete output topics
-            self.clean_app(
+            self.__clean_app(
                 values=self.to_helm_values(),
                 dry_run=dry_run,
                 delete_outputs=delete_outputs,
                 retain_clean_jobs=self.config.retain_clean_jobs,
             )
 
-            if self.to and self.handlers.schema_handler and clear_schemas:
```
Comment (philipp94831) [inline kpops/components/base_components/kafka_app.py:79]: Actually I think we can remove this whole delete_schemas for KafkaApp. We only need it for KafkaConnect

File: kpops/components/base_components/kafka_connect.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 class KafkaConnector(PipelineComponent, ABC):
     _type = "kafka-connect"
     app: KafkaConnectConfig
+    clean_schemas: bool = True
```
Comment (philipp94831) [inline kpops/components/base_components/kafka_connect.py:19]: I think we don't need this and it should always be true

File: kpops/components/base_components/kafka_app.py (hunk 2 above)
Comment (raminqaf) [reply kpops/components/base_components/kafka_app.py:79]: So it should always delete the schema?

Comment (philipp94831) [reply kpops/components/base_components/kafka_app.py:79]: No, not at all. The streams app clean up does this

File: tests/component_handlers/kafka_connect/test_connect_wrapper.py [hunk 4]
```diff
@@ -224,6 +224,7 @@ def test_should_create_correct_update_connector_request(self, mock_put: MagicMoc
         connector_name = "test-connector"
         configs = {
             "batch.size": "50",
+            "clean_schemas"
```
Comment (disrupted) [inline tests/component_handlers/kafka_connect/test_connect_wrapper.py:227]: missing comma

File: kpops/components/base_components/pipeline_component.py [hunk 5]
```diff
@@ -161,5 +161,8 @@ def inflate(self) -> list[PipelineComponent]:
     def deploy(self, dry_run: bool) -> None:
         pass
 
-    def destroy(self, dry_run: bool, clean: bool, delete_outputs: bool) -> None:
```
Comment (disrupted) [inline kpops/components/base_components/pipeline_component.py:164]:
```suggestion
    def destroy(self, clean: bool, delete_outputs: bool, dry_run: bool) -> None:
```

File: kpops/cli/main.py [hunk 6]
```diff
@@ -191,7 +191,9 @@ def run_destroy_clean_reset(
     )
     for component in reversed(get_steps_to_apply(pipeline, steps)):
         log_action("Destroy", component)
-        component.destroy(dry_run=dry_run, clean=clean, delete_outputs=delete_outputs)
+        component.destroy(dry_run=dry_run)
```
Comment (disrupted) [inline kpops/cli/main.py:194]:
```suggestion
        component.destroy(dry_run)
```

less verbose calling

**Review by disrupted — CHANGES_REQUESTED** [review]

File: kpops/cli/main.py (hunk 6 above)
Comment (disrupted) [reply kpops/cli/main.py:194]: or make `dry_run` a keyword-only argument for consistency. up to you

File: kpops/components/base_components/pipeline_component.py (hunk 5 above)
Comment (disrupted) [reply kpops/components/base_components/pipeline_component.py:164]: move `dry_run` parameter to the end so that it aligns with the other commands

**Comment by disrupted** [comment]
shouldn't there be a separate method for `reset` as well? (according to the linked issue)

File: kpops/cli/main.py [hunk 7]
```diff
@@ -173,33 +173,19 @@ def log_action(action: str, pipeline_component: PipelineComponent):
...
-    steps: str | None,
-    verbose,
-    clean: bool,
-    delete_outputs: bool,
+def get_pipeline(
+    components_module, config, defaults, pipeline_base_dir, pipeline_path, verbose
 ):
     pipeline_config = create_pipeline_config(config, defaults, verbose)
-    pipeline = setup_pipeline(
+    return setup_pipeline(
```
Comment (disrupted) [inline kpops/cli/main.py:180]: good idea to extract it 👍🏻 
why not add it to `setup_pipeline`?

File: kpops/components/base_components/kafka_app.py [hunk 8]
```diff
@@ -63,33 +67,27 @@ def deploy(self, dry_run: bool):
...
-            # producer clean up job will delete output topics
-            self.clean_app(
-                values=self.to_helm_values(),
-                dry_run=dry_run,
-                delete_outputs=delete_outputs,
-                retain_clean_jobs=self.config.retain_clean_jobs,
-            )
+    @override
+    def destroy(self, dry_run: bool) -> None:
+        super().destroy(dry_run)
```
Comment (disrupted) [inline kpops/components/base_components/kafka_app.py:72]: what is the purpose of this override?

File: kpops/components/base_components/kafka_app.py [hunk 9]
```diff
@@ -63,33 +67,27 @@ def deploy(self, dry_run: bool):
...
-                self.handlers.schema_handler.delete_schemas(self.to, dry_run)
+    @override
+    def clean(self, dry_run: bool):
+        self.clean_app(
+            values=self.to_helm_values(),
+            dry_run=dry_run,
+            retain_clean_jobs=self.config.retain_clean_jobs,
+        )
 
     def clean_app(
```
Comment (disrupted) [inline kpops/components/base_components/kafka_app.py:82]: if we're exposing the `clean` method as part of the API, this should be internal. you can prefix it using 1 or 2 underscores

File: kpops/components/base_components/kafka_app.py (hunk 8 above)
Comment (raminqaf) [reply kpops/components/base_components/kafka_app.py:72]: Ah I thought the hierarchy is important but removed it and the tests passed

File: kpops/components/base_components/kafka_app.py (hunk 9 above)
Comment (raminqaf) [reply kpops/components/base_components/kafka_app.py]: Good point I did it with 1 underscore to make it protected since the `streams_app.py` is using it

File: kpops/components/base_components/kafka_app.py [hunk 10]
```diff
@@ -63,33 +67,27 @@ def deploy(self, dry_run: bool):
...
-                retain_clean_jobs=self.config.retain_clean_jobs,
-            )
+    @override
+    def destroy(self, dry_run: bool) -> None:
+        super().destroy(dry_run)
 
-            if self.to and self.handlers.schema_handler and clear_schemas:
-                self.handlers.schema_handler.delete_schemas(self.to, dry_run)
+    @override
+    def clean(self, dry_run: bool):
```
Comment (disrupted) [inline kpops/components/base_components/kafka_app.py:71]: please add missing return types :)
```suggestion
    def clean(self, dry_run: bool) -> None:
```

File: kpops/components/base_components/kafka_app.py [hunk 11]
```diff
@@ -63,33 +67,23 @@ def deploy(self, dry_run: bool):
...
         self,
         values: dict,
         dry_run: bool,
-        delete_outputs: bool = False,
         retain_clean_jobs: bool = False,
-    ):
+    ) -> None:
         """
         Cleans an app using the respective cleanup job
         :param dry_run: Dry run command
```
Comment (disrupted) [inline kpops/components/base_components/kafka_app.py:86]: please fix the parameter order in the doc string

File: kpops/components/streams_bootstrap/streams/streams_app.py [hunk 12]
```diff
@@ -63,10 +65,25 @@ def helm_repo_config(self) -> HelmRepoConfig | None:
...
+
+    @override
+    def clean(self, dry_run: bool) -> None:
+        values = self.to_helm_values()
+        values["streams"]["deleteOutput"] = True
+        self._clean_app(
+            values=values,
+            dry_run=dry_run,
+            retain_clean_jobs=self.config.retain_clean_jobs,
+        )
```
Comment (disrupted) [inline kpops/components/streams_bootstrap/streams/streams_app.py:86]: these methods are very similar, perhaps you could extract some of it into `_clean_app`

File: kpops/cli/main.py [hunk 13]
```diff
@@ -93,8 +93,11 @@ def setup_pipeline(
     pipeline_base_dir: Path,
     pipeline_path: Path,
     components_module: str | None,
-    pipeline_config: PipelineConfig,
+    defaults: Path,
```
Comment (philipp94831) [inline kpops/cli/main.py:96]: why did you change the parameter?

File: kpops/cli/main.py [hunk 14]
```diff
@@ -270,18 +250,12 @@ def destroy(
...
-        dry_run,
-        pipeline_path,
-        steps,
-        verbose,
-        clean=False,
-        delete_outputs=False,
+    pipeline = setup_pipeline(
+        pipeline_base_dir, pipeline_path, components_module, defaults, config, verbose
     )
+    for component in reversed(get_steps_to_apply(pipeline, steps)):
```
Comment (philipp94831) [inline kpops/cli/main.py:256]: extract a method please

File: tests/components/test_producer_app.py [hunk 15]
```diff
@@ -256,16 +183,74 @@ def test_should_clean_streams_app_and_deploy_clean_up_job_and_delete_clean_up(
...
-                    True,
+                ),
+                mocker.call.helm_uninstall(
+                    "test-namespace", "example-name-clean", True
                 ),
             ]
         )
+
+    @staticmethod
+    def __create_producer_app(config: PipelineConfig, handlers: ComponentHandlers):
```
Comment (disrupted) [inline tests/components/test_producer_app.py:235]: consider making this a pytest fixture instead, if possible

File: kpops/components/base_components/kafka_app.py [hunk 16]
```diff
@@ -63,33 +67,23 @@ def deploy(self, dry_run: bool):
...
+    @override
+    def clean(self, dry_run: bool) -> None:
+        self._clean_app(
+            values=self.to_helm_values(),
+            dry_run=dry_run,
+            retain_clean_jobs=self.config.retain_clean_jobs,
+        )
 
-    def clean_app(
+    def _clean_app(
```
Comment (philipp94831) [inline kpops/components/base_components/kafka_app.py:78]: _run_clean_up_job

File: kpops/components/base_components/kafka_app.py [hunk 17]
```diff
@@ -63,33 +67,23 @@ def deploy(self, dry_run: bool):
...
-                values=self.to_helm_values(),
-                dry_run=dry_run,
-                delete_outputs=delete_outputs,
-                retain_clean_jobs=self.config.retain_clean_jobs,
-            )
-
-            if self.to and self.handlers.schema_handler and clear_schemas:
-                self.handlers.schema_handler.delete_schemas(self.to, dry_run)
+    @override
+    def clean(self, dry_run: bool) -> None:
```
Comment (philipp94831) [inline kpops/components/base_components/kafka_app.py:71]: This should move to producer

File: kpops/components/base_components/kafka_connect.py [hunk 18]
```diff
@@ -75,21 +71,22 @@ def apply_from_inputs(self, name: str, topic: FromTopic) -> NoReturn:
...
-                connector_name=self.name,
-                connector_type=KafkaConnectorType.SOURCE,
-                dry_run=dry_run,
-                retain_clean_jobs=self.config.retain_clean_jobs,
-                offset_topic=os.environ[
-                    f"{ENV_PREFIX}KAFKA_CONNECT_RESETTER_OFFSET_TOPIC"
-                ],
-            )
+    def reset(self, dry_run: bool) -> None:
+        self.__clean_source_connector(dry_run)
```
Comment (philipp94831) [inline kpops/components/base_components/kafka_connect.py:75]: the method name is probably a bit confusing because it uses the term clean

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: tests/components/test_streams_app.py [hunk 19]
```diff
@@ -41,6 +41,29 @@ def config(self) -> PipelineConfig:
             pipeline_prefix="",
         )
 
+    @pytest.fixture
+    def streams_app(self, config: PipelineConfig, handlers: ComponentHandlers):
```
Comment (disrupted) [inline tests/components/test_streams_app.py:45]:
```suggestion
    def streams_app(self, config: PipelineConfig, handlers: ComponentHandlers) -> StreamsApp:
```

File: kpops/components/base_components/kubernetes_app.py [hunk 20]
```diff
@@ -83,15 +88,16 @@ def deploy(self, dry_run: bool) -> None:
...
+    def destroy(self, dry_run: bool) -> None:
         stdout = self.helm.uninstall(
             self.namespace,
             self.helm_release_name,
             dry_run,
         )
-        if dry_run and self.helm_diff.config.enable and stdout:
-            self.print_helm_diff(stdout)
+
+        if dry_run and stdout:
```
Comment (philipp94831) [inline kpops/components/base_components/kubernetes_app.py:99]: why do you check for dry run?

**Review by philipp94831 — CHANGES_REQUESTED** [review]

**Review by philipp94831 — APPROVED** [review]
